File: scripts/skillopt-eval/contracts/run-lock.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { z } from "zod";
import {
  ArtifactIdSchema,
  CONTRACT_SCHEMA_VERSION,
  ContractIntegrityError,
  JsonValueSchema,
  NonEmptyStringSchema,
  Sha256Schema,
  boundedContractSchema,
  contractHash,
  parseJsonText,
} from "./common";
import { CodexGatesSchema } from "./gates";

const ModelPricingSchema = z
  .object({
    inputPerMillionTokens: z.number().nonnegative(),
    cachedInputPerMillionTokens: z.number().nonnegative(),
    outputPerMillionTokens: z.number().nonnegative(),
  })
  .strict();

export const SourceLockSchema = z
  .object({
    package: z.literal("skillopt"),
    version: NonEmptyStringSchema,
    commit: z.string().regex(/^[a-f0-9]{40}$/),
    repository: z.url(),
    license: NonEmptyStringSchema,
    retrievedAt: z.iso.date(),
    python: NonEmptyStringSchema,
  })
  .strict();

const DEFAULT_SOURCE_LOCK_PATH = resolve(
  import.meta.dir,
  "../../../tools/skillopt/source-lock.json",
);

// implements REQ-skillopt-codex-optimization
export function loadSkillOptSourceLock(path = DEFAULT_SOURCE_LOCK_PATH) {
  return SourceLockSchema.parse(JSON.parse(readFileSync(path, "utf8")));
}

// implements REQ-skillopt-codex-optimization
export const SKILLOPT_SOURCE_LOCK = loadSkillOptSourceLock();

const ExecutableIdentitySchema = z
  .object({
    path: NonEmptyStringSchema,
    version: NonEmptyStringSchema,
    sha256: Sha256Schema,
  })
  .strict();

const SkillSurfaceHashesSchema = z
  .object({
    bodyHash: Sha256Schema,
    frontmatterHash: Sha256Schema,
    resourcesHash: Sha256Schema,
  })
  .strict();

// implements REQ-skillopt-codex-optimization
export const PricingTableSchema = z
  .object({
    name: z.literal("price-equivalent-estimates"),
    effectiveFrom: z.iso.date(),
    currency: z.literal("USD"),
    source: NonEmptyStringSchema,
    models: z
      .object({
        "gpt-5.4-mini": ModelPricingSchema,
        "gpt-5.6-sol": ModelPricingSchema,
      })
      .strict(),
  })
  .strict();

const DirtyStateSchema = z.discriminatedUnion("isDirty", [
  z.object({ isDirty: z.literal(false), diffHash: z.null() }).strict(),
  z.object({ isDirty: z.literal(true), diffHash: Sha256Schema }).strict(),
]);

// implements REQ-skillopt-codex-optimization
export function createRunLockSchema(sourceLockPath = DEFAULT_SOURCE_LOCK_PATH) {
  loadSkillOptSourceLock(sourceLockPath);
  return boundedContractSchema(
    z
      .object({
        schemaVersion: z.literal(CONTRACT_SCHEMA_VERSION),
        artifactType: z.literal("run-lock"),
        runId: ArtifactIdSchema,
        repositoryHashAlgorithm: z.enum(["sha1", "sha256"]),
        repositoryCommit: z.string().regex(/^(?:[a-f0-9]{40}|[a-f0-9]{64})$/),
        dirtyState: DirtyStateSchema,
        codexCliVersion: NonEmptyStringSchema,
        codexExecutable: ExecutableIdentitySchema,
        cliArgs: z.array(NonEmptyStringSchema).min(1),
        artifactRoot: NonEmptyStringSchema,
        targetModel: z.literal("gpt-5.4-mini"),
        optimizerModel: z.literal("gpt-5.6-sol"),
        skillopt: z
          .object({
            package: z.literal("skillopt"),
            version: NonEmptyStringSchema,
            commit: z.string().regex(/^[a-f0-9]{40}$/),
            repository: z.url(),
            license: NonEmptyStringSchema,
            retrievedAt: z.iso.date(),
            python: NonEmptyStringSchema,
            packageHash: Sha256Schema,
            sourceHash: Sha256Schema,
            uvLockHash: Sha256Schema,
          })
          .strict(),
        sourceLockHash: Sha256Schema,
        catalogHash: Sha256Schema,
        fixtureHash: Sha256Schema,
        fixtureGeneratorHash: Sha256Schema,
        pricing: PricingTableSchema,
        pricingHash: Sha256Schema,
        baselineSkillHashes: z
          .object({
            "kibi-usage": SkillSurfaceHashesSchema,
            "kibi-freshness": SkillSurfaceHashesSchema,
            "kibi-traceability": SkillSurfaceHashesSchema,
            "kibi-bootstrap": SkillSurfaceHashesSchema,
          })
          .strict(),
        seed: z.int().nonnegative(),
        authMode: z.literal("existing-login"),
        hosts: z.tuple([z.literal("codex")]),
        gates: CodexGatesSchema,
      })
      .strict()
      .superRefine((lock, context) => {
        const expectedLength =
          lock.repositoryHashAlgorithm === "sha1" ? 40 : 64;
        if (lock.repositoryCommit.length !== expectedLength) {
          context.addIssue({
            code: "custom",
            message: `${lock.repositoryHashAlgorithm} object ID must be ${expectedLength} characters`,
            path: ["repositoryCommit"],
          });
        }
      }),
  );
}

// implements REQ-skillopt-codex-optimization
export const RunLockSchema = createRunLockSchema();

// implements REQ-skillopt-codex-optimization
export type RunLock = Readonly<z.infer<typeof RunLockSchema>>;

// implements REQ-skillopt-codex-optimization
export function parseRunLockText(
  text: string,
  sourceLockPath = DEFAULT_SOURCE_LOCK_PATH,
): RunLock {
  const lock = RunLockSchema.parse(parseJsonText(text));
  const pricing = JsonValueSchema.parse(lock.pricing);
  if (contractHash(pricing) !== lock.pricingHash) {
    throw new ContractIntegrityError("pricing hash mismatch", "pricingHash");
  }
  const sourceLock = loadSkillOptSourceLock(sourceLockPath);
  const expectedPin = JsonValueSchema.parse(sourceLock);
  const actualPin = JsonValueSchema.parse({
    package: lock.skillopt.package,
    version: lock.skillopt.version,
    commit: lock.skillopt.commit,
    repository: lock.skillopt.repository,
    license: lock.skillopt.license,
    retrievedAt: lock.skillopt.retrievedAt,
    python: lock.skillopt.python,
  });
  if (contractHash(actualPin) !== contractHash(expectedPin)) {
    throw new ContractIntegrityError("source lock mismatch", "skillopt");
  }
  if (contractHash(expectedPin) !== lock.sourceLockHash) {
    throw new ContractIntegrityError(
      "source lock hash mismatch",
      "sourceLockHash",
    );
  }
  return lock;
}

// implements REQ-skillopt-codex-optimization
export function runLockHash(lock: RunLock): string {
  return contractHash(JsonValueSchema.parse(lock));
}

// implements REQ-skillopt-codex-optimization
export function assertRunLockMatches(expected: RunLock, actual: RunLock): void {
  if (expected.dirtyState.isDirty || actual.dirtyState.isDirty) {
    throw new ContractIntegrityError("dirty run lock", "dirtyState");
  }
  assertMatchingRunLockHash(expected, actual);
}

export function assertMatchingRunLockHash(
  expected: RunLock,
  actual: RunLock,
): void {
  if (runLockHash(expected) !== runLockHash(actual)) {
    throw new ContractIntegrityError("immutable run lock mismatch", "runLock");
  }
}

File: tools/skillopt/kibi_skillopt/run_lock.py
from __future__ import annotations

from datetime import date
from pathlib import Path
from typing import Annotated, Literal

from pydantic import Field, model_validator
from typing_extensions import Self

from .common import (
    ArtifactId,
    ContractModel,
    ContractValidationError,
    JsonInteger,
    JsonNumber,
    JsonValue,
    NonEmptyString,
    Sha256,
    contract_hash,
    parse_json_value,
)


class CleanState(ContractModel):
    is_dirty: Annotated[Literal[False], Field(alias="isDirty")]
    diff_hash: Annotated[None, Field(alias="diffHash")]


class DirtyState(ContractModel):
    is_dirty: Annotated[Literal[True], Field(alias="isDirty")]
    diff_hash: Annotated[Sha256, Field(alias="diffHash")]


class ModelPricing(ContractModel):
    input_per_million_tokens: Annotated[JsonNumber, Field(alias="inputPerMillionTokens", ge=0)]
    cached_input_per_million_tokens: Annotated[
        JsonNumber, Field(alias="cachedInputPerMillionTokens", ge=0)
    ]
    output_per_million_tokens: Annotated[JsonNumber, Field(alias="outputPerMillionTokens", ge=0)]


class PricingModels(ContractModel):
    target: Annotated[ModelPricing, Field(alias="gpt-5.4-mini")]
    optimizer: Annotated[ModelPricing, Field(alias="gpt-5.6-sol")]


class PricingTable(ContractModel):
    name: Literal["price-equivalent-estimates"]
    effective_from: Annotated[date, Field(alias="effectiveFrom")]
    currency: Literal["USD"]
    source: NonEmptyString
    models: PricingModels


class SkillOptPin(ContractModel):
    package: Literal["skillopt"]
    version: NonEmptyString
    commit: Annotated[str, Field(pattern=r"^[a-f0-9]{40}$")]
    repository: NonEmptyString
    license: NonEmptyString
    retrieved_at: Annotated[date, Field(alias="retrievedAt")]
    python: NonEmptyString
    package_hash: Annotated[Sha256, Field(alias="packageHash")]
    source_hash: Annotated[Sha256, Field(alias="sourceHash")]
    uv_lock_hash: Annotated[Sha256, Field(alias="uvLockHash")]


class SkillSurfaceHashes(ContractModel):
    body_hash: Annotated[Sha256, Field(alias="bodyHash")]
    frontmatter_hash: Annotated[Sha256, Field(alias="frontmatterHash")]
    resources_hash: Annotated[Sha256, Field(alias="resourcesHash")]


class BaselineSkillHashes(ContractModel):
    usage: Annotated[SkillSurfaceHashes, Field(alias="kibi-usage")]
    freshness: Annotated[SkillSurfaceHashes, Field(alias="kibi-freshness")]
    traceability: Annotated[SkillSurfaceHashes, Field(alias="kibi-traceability")]
    init: Annotated[SkillSurfaceHashes, Field(alias="init-kibi")]


class ExecutableIdentity(ContractModel):
    path: NonEmptyString
    version: NonEmptyString
    sha256: Sha256


class SourceLock(ContractModel):
    package: Literal["skillopt"]
    version: NonEmptyString
    commit: Annotated[str, Field(pattern=r"^[a-f0-9]{40}$")]
    repository: NonEmptyString
    license: NonEmptyString
    retrieved_at: Annotated[date, Field(alias="retrievedAt")]
    python: NonEmptyString


DEFAULT_SOURCE_LOCK_PATH = Path(__file__).resolve().parents[1] / "source-lock.json"


def load_skillopt_source_lock(path: Path = DEFAULT_SOURCE_LOCK_PATH) -> SourceLock:
    return SourceLock.model_validate_json(path.read_text(encoding="utf-8"))


SKILLOPT_SOURCE_LOCK = load_skillopt_source_lock()


class CandidateMeanDelta(ContractModel):
    baseline: Literal[8]
    one_shot: Annotated[Literal[5], Field(alias="oneShot")]


class CandidateHardPassDelta(ContractModel):
    baseline: Literal[2]
    one_shot: Annotated[Literal[1], Field(alias="oneShot")]


class CandidateGates(ContractModel):
    mean_minimum: Annotated[Literal[85], Field(alias="meanMinimum")]
    hard_passes_minimum: Annotated[Literal[13], Field(alias="hardPassesMinimum")]
    hard_passes_total: Annotated[Literal[16], Field(alias="hardPassesTotal")]
    mean_delta_minimum: Annotated[CandidateMeanDelta, Field(alias="meanDeltaMinimum")]
    hard_pass_delta_minimum: Annotated[CandidateHardPassDelta, Field(alias="hardPassDeltaMinimum")]


class BootstrapGates(ContractModel):
    resamples: Literal[10000]
    seed: Literal[5417]
    confidence_level: Annotated[JsonNumber, Field(alias="confidenceLevel", ge=0.95, le=0.95)]
    sidedness: Literal["one-sided"]
    lower_bound_exclusive_minimum: Annotated[Literal[0], Field(alias="lowerBoundExclusiveMinimum")]
    cluster_unit: Annotated[Literal["task"], Field(alias="clusterUnit")]


class FamilyGuard(ContractModel):
    max_mean_regression: Annotated[Literal[3], Field(alias="maxMeanRegression")]
    max_hard_pass_regression: Annotated[Literal[1], Field(alias="maxHardPassRegression")]


class BundleMeanDelta(ContractModel):
    baseline: Literal[3]
    one_shot: Annotated[Literal[3], Field(alias="oneShot")]


class BundleGates(ContractModel):
    mean_minimum: Annotated[Literal[85], Field(alias="meanMinimum")]
    hard_passes_minimum: Annotated[Literal[7], Field(alias="hardPassesMinimum")]
    hard_passes_total: Annotated[Literal[8], Field(alias="hardPassesTotal")]
    mean_delta_minimum: Annotated[BundleMeanDelta, Field(alias="meanDeltaMinimum")]
    allow_hard_pass_loss: Annotated[Literal[False], Field(alias="allowHardPassLoss")]
    max_critical_failures: Annotated[Literal[0], Field(alias="maxCriticalFailures")]


class CodexGates(ContractModel):
    held_out_tasks_per_variant: Annotated[Literal[16], Field(alias="heldOutTasksPerVariant")]
    family_slices: Annotated[Literal[4], Field(alias="familySlices")]
    bundle_tasks: Annotated[Literal[8], Field(alias="bundleTasks")]
    candidate: CandidateGates
    bootstrap: BootstrapGates
    family_guard: Annotated[FamilyGuard, Field(alias="familyGuard")]
    bundle: BundleGates


class RunLock(ContractModel):
    schema_version: Annotated[Literal["1.0.0"], Field(alias="schemaVersion")]
    artifact_type: Annotated[Literal["run-lock"], Field(alias="artifactType")]
    run_id: Annotated[ArtifactId, Field(alias="runId")]
    repository_hash_algorithm: Annotated[
        Literal["sha1", "sha256"], Field(alias="repositoryHashAlgorithm")
    ]
    repository_commit: Annotated[
        str, Field(alias="repositoryCommit", pattern=r"^(?:[a-f0-9]{40}|[a-f0-9]{64})$")
    ]
    dirty_state: Annotated[
        Annotated[CleanState | DirtyState, Field(discriminator="is_dirty")],
        Field(alias="dirtyState"),
    ]
    codex_cli_version: Annotated[NonEmptyString, Field(alias="codexCliVersion")]
    codex_executable: Annotated[ExecutableIdentity, Field(alias="codexExecutable")]
    cli_args: Annotated[list[NonEmptyString], Field(alias="cliArgs", min_length=1)]
    artifact_root: Annotated[NonEmptyString, Field(alias="artifactRoot")]
    target_model: Annotated[Literal["gpt-5.4-mini"], Field(alias="targetModel")]
    optimizer_model: Annotated[Literal["gpt-5.6-sol"], Field(alias="optimizerModel")]
    skillopt: SkillOptPin
    source_lock_hash: Annotated[Sha256, Field(alias="sourceLockHash")]
    catalog_hash: Annotated[Sha256, Field(alias="catalogHash")]
    fixture_hash: Annotated[Sha256, Field(alias="fixtureHash")]
    fixture_generator_hash: Annotated[Sha256, Field(alias="fixtureGeneratorHash")]
    pricing: PricingTable
    pricing_hash: Annotated[Sha256, Field(alias="pricingHash")]
    baseline_skill_hashes: Annotated[BaselineSkillHashes, Field(alias="baselineSkillHashes")]
    seed: Annotated[JsonInteger, Field(ge=0)]
    auth_mode: Annotated[Literal["existing-login"], Field(alias="authMode")]
    hosts: tuple[Literal["codex"]]
    gates: CodexGates

    @model_validator(mode="after")
    def verify_repository_object_id(self) -> Self:
        expected_length = 40 if self.repository_hash_algorithm == "sha1" else 64
        if len(self.repository_commit) != expected_length:
            raise ContractValidationError(
                f"{self.repository_hash_algorithm} object ID must be {expected_length} characters"
            )
        return self

    def verify_integrity(self, source_lock_path: Path) -> None:
        pricing = parse_json_value(self.pricing.model_dump_json(by_alias=True))
        if contract_hash(pricing) != self.pricing_hash:
            raise ContractValidationError("pricing hash mismatch")
        self.assert_source_lock(source_lock_path)

    def assert_source_lock(self, source_lock_path: Path) -> None:
        source_lock = load_skillopt_source_lock(source_lock_path)
        expected_pin = source_lock.model_dump(mode="json", by_alias=True)
        actual_pin = self.skillopt.model_dump(
            mode="json",
            by_alias=True,
            exclude={"package_hash", "source_hash", "uv_lock_hash"},
        )
        if actual_pin != expected_pin:
            raise ContractValidationError("SkillOpt source lock mismatch")
        source_lock_value = parse_json_value(source_lock.model_dump_json(by_alias=True))
        if contract_hash(source_lock_value) != self.source_lock_hash:
            raise ContractValidationError("SkillOpt source lock hash mismatch")

    @classmethod
    def model_validate_with_source_lock(cls, value: JsonValue, source_lock_path: Path) -> Self:
        lock = cls.model_validate(value)
        lock.verify_integrity(source_lock_path)
        return lock


def run_lock_hash(lock: RunLock) -> str:
    value = parse_json_value(lock.model_dump_json(by_alias=True))
    return contract_hash(value)


def assert_run_lock_matches(expected: RunLock, actual: RunLock) -> None:
    if expected.dirty_state.is_dirty or actual.dirty_state.is_dirty:
        raise ContractValidationError("dirty run lock")
    if run_lock_hash(expected) != run_lock_hash(actual):
        raise ContractValidationError("immutable run lock mismatch")
